Fix ignoring of missing output files

harry-wood opened the pull request:
Add a test of the 'missing output file' situation (fails running on ruby versions >2.0.0)

Explicitly test File.exists? rather than catching and quashing a 'No such file or directory' exception (Fixes it for versions >2.0.0)

Fixes https://opencorporates.atlassian.net/browse/OCD-1011

Commit:
Add a test for missing output files

We've noticed that the implementation which handle this doesn't work with ruby 2.1 and later (because the implementation relies on the text in an exception message and the text changed in that version).

https://opencorporates.atlassian.net/browse/OCD-1011

Commit:
Fix the ignoring of missing output files

Explicitly test File.exists? rather than catching and quashing a 'No such file or directory' exception (the text of which changed in ruby 2.1).

Fixes https://opencorporates.atlassian.net/browse/OCD-1011.

bensymonds commented on spec/lib/runner_spec.rb:
```diff
@@ -362,6 +362,16 @@ def handle_valid_record(record, data_type)
+        runner = test_runner('bot-with-transformer',
+          :record_handler => @handler
+        )
+        FileUtils.rm_rf(File.join(runner.base_directory, "output"))
```
I was wondering why this was necessary given the runner hadn't actually been run (ie `test_runner('bot-with-transformer').run` hadn't been called) as it is in other specs. On closer inspection it seems it _shouldn't_ be necessary, except the _other specs_ are sometimes (depending on the order specs are run in) creating the directory and output files. Obviously this is a bad thing because tests are meant to be isolated. We should look at fixing that first, which would then simplify this spec.

harry-wood replied:
Yes at the moment this test is ensuring the files are non-existant. We could make sure all the other tests are undoing their file changes in an "after" block. Would that be best practice? or we could do _both_ perhaps.

harry-wood replied:
OK. I've added lots of "after" blocks because each test should clean up any files it creates. This is repetitive, and it's tempting to reduce repetition by assuming the setting of `@runner`, but maybe that would be a bit global-variable-ish

bensymonds commented on spec/lib/runner_spec.rb:
```diff
@@ -369,6 +443,10 @@ def handle_valid_record(record, data_type)
 
+    after do
+      cleanup_test_runner(@runner)
+    end
```
It's a bit repetitive isn't it! I guess you can't simply put at top of file because there are some specs that don't have a `@runner`. But maybe those specs could be tweaked so they do..?

harry-wood replied:
ok yes. I have put made all specs use `@runner` consistently, and then put this in an after block at the top instead. ✅

bensymonds reviewed (changes requested):
In 145c344529c320dd47af454779c721f3ad04cebc you also seem to have bumped the `schema` submodule. Was that accidental? If so can you revert it?

bensymonds reviewed (changes requested):
On further inspection it seems like you actually *un*bumped the schema, from https://github.com/openc/openc-schema/commit/857dd68929fbd776294382ebc1399ec2538ed08b back to https://github.com/openc/openc-schema/commit/ff8e57ec89d5788eb07fe09cbf3b6e2d67f79e67! How did you manage that?? :smile: So definitely revert it!

Commit:
Use runner instance variables and before blocks consistently

Where possible change specs to make use of a `@runner` instance variable for consistency (and to allow generic 'after' block handling of this variable in future).

Commit:
Add after block to cleanup test files

Every spec test should clean up after itself to avoid interfering with other specs.

There are a few specs which don't use the instance variable because they need to re-initialise the bot with a different handler.

harry-wood commented:
hmm yes. I managed to tangle with the schema submodule there previously, hopefully that's all fixed.

bensymonds committed:
Add/clarify some spec contexts

And use them to share setup.

bensymonds commented:
@harry-wood Take a look at this now. I think I'm ok with it...

harry-wood commented:
Looks good.

I think I asked this before but, d'you have any preference between:
* Separate commits for adding a new test and then getting that test to pass. Represents a test-driven-development order of events.
* Squash those two commits, so test and fix are together. This way there is no point in the commit history where the tests are failing.

bensymonds commented:
Yeah good question. I'm a bit on the fence with this one. Most of the time I think I would do the latter, but probably in the past I have done the former. I'm not really bothered about having a failing commit in the history, it's more just about what's clearer to a person reading it later. Right now (I reserve the right to change my mind of course 😎) I'm thinking the latter is clearer because the commit can just be saying "Fix bug in blah blah" and can mention there was no test coverage for it. The former would have to have the first commit saying "Add failing spec for blah blah" and then surely in the description would have to say/imply there is a fix coming in the next commit..? Otherwise it's like you're adding a failing spec but then leaving it there, which would never happen (because then the specs would be broken in master).
